src/lib/network_pool/snark_pool.ml

```diff
@@ -0,0 +1,445 @@
+    -> incoming_diffs:Resource_pool.Diff.t Envelope.Incoming.t
+                      Linear_pipe.Reader.t
+    -> frontier_broadcast_pipe:transition_frontier option
+                               Broadcast_pipe.Reader.t
+    -> t Deferred.t
+
+  val add_completed_work :
+       t
+    -> ( ( transaction_snark_statement
+         , 'b
+         , 'c
+         , 'd )
+         Snark_work_lib.Work.Single.Spec.t
+         Snark_work_lib.Work.Spec.t
+       , ledger_proof )
+       Snark_work_lib.Work.Result.t
+    -> unit Deferred.t
+end
+
+module type Transition_frontier_intf = sig
+  type work
+
+  type t
+
+  module Extensions : sig
+    module Work : sig
+      type t = work [@@deriving sexp]
+
+      module Stable : sig
+        module V1 : sig
+          type nonrec t = t [@@deriving sexp, bin_io]
+
+          include Hashable.S_binable with type t := t
+        end
+      end
+
+      include Hashable.S with type t := t
+    end
+  end
+
+  val snark_pool_refcount_pipe :
+    t -> (int * int Extensions.Work.Table.t) Pipe_lib.Broadcast_pipe.Reader.t
+end
+
+module Make (Ledger_proof : sig
+  type t [@@deriving bin_io, sexp, yojson, version]
+end) (Transaction_snark : sig
+  module Statement : sig
+    type t
+  end
+end) (Transaction_snark_work : sig
+  type t =
+    { fee: Currency.Fee.t
+    ; proofs: Ledger_proof.t list
+    ; prover: Public_key.Compressed.t }
+
+  module Statement : sig
+    type t = Transaction_snark.Statement.t list [@@deriving sexp]
+
+    module Stable :
+      sig
+        module V1 : sig
+          type t [@@deriving sexp, bin_io, yojson, version]
+
+          include Hashable.S_binable with type t := t
+        end
+      end
+      with type V1.t = t
+
+    include Hashable.S with type t := t
+  end
+
+  module Checked :
+    sig
+      type unchecked
+
+      type t
+
+      val create_unsafe : unchecked -> t
+    end
+    with type unchecked := t
+end)
+(Transition_frontier : Transition_frontier_intf
+                       with type work := Transaction_snark_work.Statement.t) :
+  S
+  with type transaction_snark_statement := Transaction_snark.Statement.t
+   and type transaction_snark_work_statement :=
+              Transaction_snark_work.Statement.t
+   and type transaction_snark_work_checked := Transaction_snark_work.Checked.t
+   and type transition_frontier := Transition_frontier.t
+   and type ledger_proof := Ledger_proof.t = struct
+  module Statement_table = Transaction_snark_work.Statement.Stable.V1.Table
+
+  module Resource_pool = struct
+    module T = struct
+      (* TODO : Version this type *)
+      type t =
+        { snark_table:
+            Ledger_proof.t list Priced_proof.Stable.V1.t Statement_table.t
+        ; mutable ref_table: int Statement_table.t option }
+      [@@deriving sexp, bin_io]
+
+      (* shadow generated bin_io code so that ref table is always None when written *)
+
+      let bin_write_t buf ~pos t =
+        let t_no_ref_tbl = {t with ref_table= None} in
```

Comment:
Can you explain a little bit more why this is always None? I got a bit confused about it.